grype/db/v6/severity.go

```diff
@@ -0,0 +1,187 @@
+	gocvss40 "github.com/pandatix/go-cvss/40"
+
+	"github.com/anchore/grype/grype/vulnerability"
+	"github.com/anchore/grype/internal/log"
+)
+
+func extractSeverities(vuln *VulnerabilityHandle) (vulnerability.Severity, []vulnerability.Cvss, error) {
+	if vuln.BlobValue == nil {
+		return vulnerability.UnknownSeverity, nil, nil
+	}
+	sev := vulnerability.UnknownSeverity
+	if len(vuln.BlobValue.Severities) > 0 {
+		var err error
+		sev, err = extractSeverity(vuln.BlobValue.Severities[0].Value)
+		if err != nil {
+			return vulnerability.UnknownSeverity, nil, fmt.Errorf("unable to extract severity: %w", err)
+		}
+	}
+	return sev, toCvss(vuln.BlobValue.Severities...), nil
+}
+
+func extractSeverity(severity any) (vulnerability.Severity, error) {
+	switch sev := severity.(type) {
+	case string:
+		return vulnerability.ParseSeverity(sev), nil
+	case CVSSSeverity:
+		metrics, err := parseCVSS(sev.Vector)
+		if err != nil {
+			return vulnerability.UnknownSeverity, fmt.Errorf("unable to parse CVSS vector: %w", err)
+		}
+		if metrics == nil {
+			return vulnerability.UnknownSeverity, nil
+		}
+		return interpretCVSS(metrics.BaseScore, sev.Version), nil
+	default:
+		return vulnerability.UnknownSeverity, nil
+	}
+}
+
+func parseCVSS(vector string) (*vulnerability.CvssMetrics, error) {
+	switch {
+	case strings.HasPrefix(vector, "CVSS:3.0"):
+		cvss, err := gocvss30.ParseVector(vector)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse CVSS v3 vector: %w", err)
+		}
+		ex := roundScore(cvss.Exploitability())
+		im := roundScore(cvss.Impact())
+		return &vulnerability.CvssMetrics{
+			BaseScore:           roundScore(cvss.BaseScore()),
+			ExploitabilityScore: &ex,
+			ImpactScore:         &im,
+		}, nil
+	case strings.HasPrefix(vector, "CVSS:3.1"):
+		cvss, err := gocvss31.ParseVector(vector)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse CVSS v3.1 vector: %w", err)
+		}
+		ex := roundScore(cvss.Exploitability())
+		im := roundScore(cvss.Impact())
+		return &vulnerability.CvssMetrics{
+			BaseScore:           roundScore(cvss.BaseScore()),
+			ExploitabilityScore: &ex,
+			ImpactScore:         &im,
+		}, nil
+	case strings.HasPrefix(vector, "CVSS:4.0"):
+		cvss, err := gocvss40.ParseVector(vector)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse CVSS v4.0 vector: %w", err)
+		}
+		// there are no exploitability and impact scores in CVSS v4.0
+		return &vulnerability.CvssMetrics{
+			BaseScore: roundScore(cvss.Score()),
+		}, nil
+	default:
+		// should be CVSS v2.0 or is invalid
+		cvss, err := gocvss20.ParseVector(vector)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse CVSS v2 vector: %w", err)
+		}
+		ex := roundScore(cvss.Exploitability())
+		im := roundScore(cvss.Impact())
+		return &vulnerability.CvssMetrics{
+			BaseScore:           roundScore(cvss.BaseScore()),
+			ExploitabilityScore: &ex,
+			ImpactScore:         &im,
+		}, nil
+	}
+}
+
+func roundScore(score float64) float64 {
```

Comment:
Is there a reason to round the score internally and not just at display time? For example: if `4.0` was where "medium" started, but the score was `3.96`, it would be "low", but get rounded up to "medium"; this doesn't seem accurate to me... but then if the score was rounded up at display time, I may be confused why it says 4.0 but it's classified as low. I feel like a truncation to 1 decimal point during display time may be the "most correct" thing to do